Linux support

dscreve opened the pull request:
This PR add Linux support and maybe Windows (not tested).

beatt83 commented on Sources/JSONWebAlgorithms/KeyManagement/KeyRepresentable/KeyRepresentable.swift:
```diff
@@ -79,10 +79,4 @@ extension secp256k1.KeyAgreement.PublicKey: KeyRepresentable {
 extension SymmetricKey: KeyRepresentable {
     public var jwk: JWK { self.jwkRepresentation }
 }
-extension SecKey: KeyRepresentable {
-    public var jwk: JWK {
-        get throws {
-            try SecKeyExtended(secKey: self).jwk()
-        }
-    }
-}
```
`SecKey` is still a big part of Apple systems, this conformance helps with the adoption of the library.

Instead of deleting the code you need to do this when SecKey is not available (there are a few other places on this PR that this is applicable):

```
#if canImport(Security)

extension SecKey: KeyRepresentable {
    public var jwk: JWK {
        get throws {
            try SecKeyExtended(secKey: self).jwk()
        }
    }
}

#else
```

dscreve replied:
I will fix this

beatt83 reviewed (changes requested):
The tests need to be updated they are failing building. This just happens because if you removed the `SecKey: KeyRepresentable` conformance, if you add the `#if canImport(Security)` and not remove the conformances the tests don't need to be fixed to run on Apple platforms.

Even so be aware that the a few tests are failing mainly the `RSA` example tests.

dscreve commented on Sources/JSONWebAlgorithms/CryptoImplementation/SecureRandom/SecureRandom.swift:
```diff
@@ -28,23 +28,29 @@ public struct SecureRandom {
+            }
+        
+    #else
+            return Data((0 ..< count).map { _ in UInt8.random(in: UInt8.min ... UInt8.max) })
```
/dev/urandom is available on most distribution  : I think we can rely on it because if this not present, this a really customised distribution

beatt83 replied:
Cool agree with that.

beatt83 commented:
Hey @dscreve . Ive been checking this and I fixed most of the problems and added a linux workflow so it can be tested, I've also tested on a docker and everything seems to run fine. Can you give me permissions so I can create a PR on your forked repo?

There is just as far as I see a remaining issue, the `SWCompression` fails on decompressing the tests from the RFC. So this library cannot be used.

beatt83 commented:
- Is there something I have to do to allow you to fork ?

Since it is a forked repo if you don't mind just on the settings allow me to push. I created a different branch on top of your main I push and you decide.

The RSA tests were also failing the way you changed, I added a ci workflow to run the tests in Linux, I want to check what is the problem if there is other way to fix it without using Security framework.

dscreve commented:
The RSA failed because you use 1024 RSA keys which is too short. Use 2048 keys and it will work.David ScrèveLe 6 mars 2025 à 19:11, Gonçalo Frade ***@***.***> a écrit :﻿
  
Is there something I have to do to allow you to fork ?

Since it is a forked repo if you don't mind just on the settings allow me to push. I created a different branch on top of your main I push and you decide.
The RSA tests were also failing the way you changed, I added a ci workflow to run the tests in Linux, I want to check what is the problem if there is other way to fix it without using Security framework.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



beatt83 left a comment (beatt83/jose-swift#40)

Is there something I have to do to allow you to fork ?

Since it is a forked repo if you don't mind just on the settings allow me to push. I created a different branch on top of your main I push and you decide.
The RSA tests were also failing the way you changed, I added a ci workflow to run the tests in Linux, I want to check what is the problem if there is other way to fix it without using Security framework.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

beatt83 commented:
@dscreve went to check this statement:

> The RSA failed because you use 1024 RSA keys which is too short. Use 2048 keys and it will work.

It is true but you can use the other init that will accept 1024 `_RSA.Encryption.PublicKey(unsafeDERRepresentation:)` and it will work I changed this on the changes I will push ;)

So just the if the `SWCompression` issue is fixed and all tests run on linux we can merge ;)

dscreve commented:
I will work on this tomorrow.David ScrèveLe 6 mars 2025 à 20:46, Gonçalo Frade ***@***.***> a écrit :﻿
  @dscreve went to check this statement:

The RSA failed because you use 1024 RSA keys which is too short. Use 2048 keys and it will work.

It is true but you can use the other init that will accept 1024 _RSA.Encryption.PublicKey(unsafeDERRepresentation:) and it will work I changed this on the changes I will push ;)
So just the if the SWCompression issue is fixed and all tests run on linux we can merge ;)—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



beatt83 left a comment (beatt83/jose-swift#40)
@dscreve went to check this statement:

The RSA failed because you use 1024 RSA keys which is too short. Use 2048 keys and it will work.

It is true but you can use the other init that will accept 1024 _RSA.Encryption.PublicKey(unsafeDERRepresentation:) and it will work I changed this on the changes I will push ;)
So just the if the SWCompression issue is fixed and all tests run on linux we can merge ;)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

dscreve commented:
I have rewritten zlib compression to be compatible with NSData implementation.
This should be ok for now.

beatt83 commented:
> I have rewritten zlib compression to be compatible with NSData implementation. This should be ok for now.

Hey mate. 
I was checking this in the weekend I left a PR for you on that repo :) .

Also can you give me the push permissions on your fork branch? I would really like to commit my changes on your PR, so we can have the GitHub actions as well.

dscreve commented:
How can I do that ? (GitHub PR is a nightmare for me…)David ScrèveLe 8 mars 2025 à 20:07, Gonçalo Frade ***@***.***> a écrit :﻿
  
I have rewritten zlib compression to be compatible with NSData implementation. This should be ok for now.

Hey mate.
I was checking this in the weekend I left a PR for you on that repo :) .
Also can you give me the push permissions on your fork branch? I would really like to commit my changes on your PR, so we can have the GitHub actions as well.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



beatt83 left a comment (beatt83/jose-swift#40)

I have rewritten zlib compression to be compatible with NSData implementation. This should be ok for now.

Hey mate.
I was checking this in the weekend I left a PR for you on that repo :) .
Also can you give me the push permissions on your fork branch? I would really like to commit my changes on your PR, so we can have the GitHub actions as well.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

beatt83 commented:
@dscreve you can go to your forked repository -> settings -> collaborators -> add me as a collaborator `beatt83`.

This will enable me to push a branch and create a pr.

Since its a fork repo there is no problem I guess ;)

<img width="1821" alt="Screenshot 2025-03-08 at 20 52 37" src="https://github.com/user-attachments/assets/7e323fc4-70cf-4dc9-b993-f4a753c78ceb" />

dscreve commented:
Ok. I’ll do it tomorrow (it’s 22PM here).David ScrèveLe 8 mars 2025 à 21:55, Gonçalo Frade ***@***.***> a écrit :﻿
  @dscreve you can go to your forked repository -> settings -> collaborators -> add me as a collaborator beatt83.
This will enable me to push a branch and create a pr.
Since its a fork repo there is no problem I guess ;)
Screenshot.2025-03-08.at.20.52.37.png (view on web)—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



beatt83 left a comment (beatt83/jose-swift#40)
@dscreve you can go to your forked repository -> settings -> collaborators -> add me as a collaborator beatt83.
This will enable me to push a branch and create a pr.
Since its a fork repo there is no problem I guess ;)
Screenshot.2025-03-08.at.20.52.37.png (view on web)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

beatt83 commented:
> Ok. I’ll do it tomorrow (it’s 22PM here)

Same :) U will be off site for a week, so I guess you got time, if I have access during the morning I will do it then ;) .

dscreve commented:
Done (also work from phone…)David ScrèveLe 8 mars 2025 à 21:58, David Scrève ***@***.***> a écrit :﻿Ok. I’ll do it tomorrow (it’s 22PM here).David ScrèveLe 8 mars 2025 à 21:55, Gonçalo Frade ***@***.***> a écrit :﻿
  @dscreve you can go to your forked repository -> settings -> collaborators -> add me as a collaborator beatt83.
This will enable me to push a branch and create a pr.
Since its a fork repo there is no problem I guess ;)
Screenshot.2025-03-08.at.20.52.37.png (view on web)—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



beatt83 left a comment (beatt83/jose-swift#40)
@dscreve you can go to your forked repository -> settings -> collaborators -> add me as a collaborator beatt83.
This will enable me to push a branch and create a pr.
Since its a fork repo there is no problem I guess ;)
Screenshot.2025-03-08.at.20.52.37.png (view on web)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

beatt83 commented:
@dscreve I still cannot push but for the sake of time I went for a different approach. I pushed here and created a new PR https://github.com/beatt83/jose-swift/pull/42 also I added you as a collaborator, hope you dont mind. ;)

beatt83 commented:
Closing this PR since we now have https://github.com/beatt83/jose-swift/pull/42 .
